src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,141 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  // 2. the base operand is the same as the current con, i.e., a + b
+  // 3. input is an already-optimized multiplication, i.e., a*6 => (a<<1) + (a<<2)
+  // 4. is simple additions not worth transforming, i.e., a + a
+  if (base == nullptr
+      || base == this
+      || is_optimized_multiplication(this, base)
+      || (base == in(1) && base == in(2))) {
+    return nullptr;
+  }
+
+  Node* con = (bt == T_INT) ? (Node*) phase->intcon((jint) factor) : (Node*) phase->longcon(factor);
+  Node* mul = MulNode::make(base, con, bt);
+
+  PhaseIterGVN* igvn = phase->is_IterGVN();
+  if (igvn != nullptr) {
+    mul = igvn->register_new_node_with_optimizer(mul);
+  }
+
+  return mul;
+}
+
+// MulINode::Ideal() optimizes a multiplication to an addition of at most two terms if possible.
+bool AddNode::is_optimized_multiplication(Node* node, Node* base) {
+  // Look for pattern: LShiftNode(a, CON)
+  if (node->is_LShift() && node->in(2)->is_Con()) {
+    return base == node->isa_LShift()->in(1);
+  }
+
+  // Look for patterns:
+  //     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+  //     - AddNode(LShiftNode(a, CON), LShiftNode(a, CON)/a)
+  // giving that lhs is different from rhs
+  if (node->is_Add()
+      && node->in(1) != node->in(2)
+      && (
+          (node->in(1)->is_LShift() && node->in(1)->in(2)->is_Con()) // AddNode(LShiftNode(a, CON), *)
+              || (node->in(2)->is_LShift() && node->in(2)->in(2)->is_Con()) // AddNode(*, LShiftNode(a, CON))
+      )) {
+    Node* a1 = node->in(1)->is_LShift() ? node->in(1)->in(1) : node->in(1);
+    Node* a2 = node->in(2)->is_LShift() ? node->in(2)->in(1) : node->in(2);
+
+    return a1 == a2 && base == a1;
+  }
+
+  // Look for pattern: SubNode(LShiftNode(a, CON), a)
+  if (node->is_Sub() && node->in(1)->is_LShift() && node->in(1)->in(2)->is_Con()) {
+    Node* a1 = node->in(1)->in(1);
+    Node* a2 = node->in(2);
+
+    return a1 == a2 && base == a1;
+  }
+
+  return false;
+}
+
+// For a series of additions, extract the base operand and the multiplier. E.g., extract a*n from a + a + ... + a
+jlong AddNode::extract_base_operand_from_serial_additions(PhaseGVN* phase, Node* node, Node** base, int depth_limit) {
+  *base = node;
+
+  if (depth_limit == 0) {
+    return 1;
+  }
+
+  // MulNode(any, const), e.g., a*2
+  if (node->is_Mul()
+      && node->Opcode() != Op_AndI && node->Opcode() != Op_AndL // AndNode extends MulNode for some reason
+      && (node->in(1)->is_Con() || node->in(2)->is_Con())) {
+    Node* const_node = node->in(1)->is_Con() ? node->in(1) : node->in(2);
+    Node* operand_node = node->in(1)->is_Con() ? node->in(2) : node->in(1);
+    BasicType bt = phase->type(const_node)->basic_type();
+
+    if (bt == T_INT || bt == T_LONG) { // const could potentially be void type
+      Node* mul_base;
+      jlong multiplier = extract_base_operand_from_serial_additions(phase, operand_node, &mul_base, depth_limit - 1);
```

Comment:
Do you need to recurse at all here?

Reply:
I believe so. Consider the case `(a + a) * 3`. Recurse here allows us to extract `a` and factor `2 * 3 => 6`

Reply:
That case would be better handled in 2 steps, I think:
`a+a` into `a*2` with a `AddNode` transformation
`(a*2)*3` into `a*6` with a `MulNode` (or `LShift`) transformation. Can you check if it already exists?

Reply:
Done. The new version doesn't use recurssions at all